### PR TITLE
8283726: x86 intrinsics for compare method in Integer and Long

### PR DESCRIPTION
**Pull request by vamsi-parasa**
Implements x86 intrinsics for compare() method in java.lang.Integer and java.lang.Long.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [ ] Change must be properly reviewed

### Issue
 * [JDK-8283726](https://bugs.openjdk.java.net/browse/JDK-8283726): x86 intrinsics for compare method in Integer and Long

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/7975/head:pull/7975` \
`$ git checkout pull/7975`

Update a local copy of the PR: \
`$ git checkout pull/7975` \
`$ git pull https://git.openjdk.java.net/jdk pull/7975/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 7975`

View PR using the GUI difftool: \
`$ git pr show -t 7975`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/7975.diff">https://git.openjdk.java.net/jdk/pull/7975.diff</a>

</details>

### CODE REVIEW

**Comment by merykitty** [comment]
This is both complicated and inefficient, I would suggest building the intrinsic in the IR graph so that the compiler can simplify `Integer.compareUnsigned(x, y) < 0` into `x u< y`. Thanks.

File: test/micro/org/openjdk/bench/java/lang/CompareInteger.java [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+
+            if (!mode.equals("mixed")) {
+                boolean doSwap = (mode.equals("lessThanEqual") && input1[i] > input2[i]) ||
+                                (mode.equals("greaterThanEqual") && input1[i] < input2[i]);
+                if (doSwap) {
+                    int tmp = input1[i];
+                    input1[i] = input2[i];
+                    input2[i] = tmp;
+                }
+            }
```
Comment (jatin-bhateja) [inline test/micro/org/openjdk/bench/java/lang/CompareInteger.java:78]: Logic re-organization suggestion:- 
 
```
 for (int i = 0 ; i < BUFFER_SIZE; i++) {
    input1[i] = rng.nextLong();
 }

 if (mode.equals("equals") {
    GRADIANT = 0;
 } else if (mode.equals("greaterThanEquals")) {
    GRADIANT = 1;
 } else {
    assert mode.equals("lessThanEqual");
    GRADIANT = -1;
 }

 for(int i = 0 ; i < BUFFER_SIZE; i++) {
    input2[i] = input1[i] + i*GRADIANT;
 }
```

File: test/micro/org/openjdk/bench/java/lang/CompareLong.java [hunk 2]
```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright (c) 2022, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
```
Comment (jatin-bhateja) [inline test/micro/org/openjdk/bench/java/lang/CompareLong.java:5]: We can unify this benchmark along with integer compare micro.

File: src/hotspot/cpu/x86/x86_64.ad [hunk 3]
```diff
@@ -12100,6 +12100,112 @@ instruct overflowMulL_rReg_imm(rFlagsReg cr, rRegL op1, immL32 op2, rRegL tmp)
   ins_pipe(ialu_reg_reg_alu0);
 %}
 
+//----------Intrinsics for Compare function------------------------------------
+
+instruct compareSignedI_rReg(rRegI dst, rRegI op1, rRegI op2, rRegI tmp, rFlagsReg cr)
+%{
+  match(Set dst (CompareSignedI op1 op2));
```
Comment (jatin-bhateja) [inline src/hotspot/cpu/x86/x86_64.ad:12107]: Please also include these patterns in x86_32.ad

File: src/hotspot/cpu/x86/x86_64.ad [hunk 4]
```diff
@@ -12100,6 +12100,112 @@ instruct overflowMulL_rReg_imm(rFlagsReg cr, rRegL op1, immL32 op2, rRegL tmp)
...
+
+    Label done;
+    __ movl(tmp, -1);
+    __ cmpl(op1, op2);
+    __ jccb(Assembler::less, done);
+	  __ movl(tmp, 1);
+    __ jccb(Assembler::greater, done);
+	  __ movl(tmp, 0);
+    __ bind(done);
+    __ movl(dest, tmp);
```
Comment (jatin-bhateja) [inline src/hotspot/cpu/x86/x86_64.ad:12125]: Please move this in macro-assembly routine.

**Comment by vamsi-parasa** [comment]
> This is both complicated and inefficient, I would suggest building the intrinsic in the IR graph so that the compiler can simplify `Integer.compareUnsigned(x, y) < 0` into `x u< y`. Thanks.

Thank you the suggestion! This intrinsic uses 1 cmp instruction instead of two and shows ~10% improvement due to better branch prediction. Even without the intrinsic, the compiler is currently able to reduce it to x u< y but is still generating two cmp (unsigned) instructions as Integer.compareUnsigned(x, y) is implemented as x u< y? -1 : (x ==y ? 0: 1)

File: src/hotspot/share/opto/comparenode.hpp [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+    CompareUnsignedINode(Node* in1, Node* in2) : CompareNode(in1, in2) {}
+    virtual int Opcode() const;
+};
+
+//---------- CompareUnsignedLNode -----------------------------------------------------
+class CompareUnsignedLNode : public CompareNode {
+    public:
+    CompareUnsignedLNode(Node* in1, Node* in2) : CompareNode(in1, in2) {}
+    virtual int Opcode() const;
+};
```
[the remark this answers is not in the document]
Comment (vamsi-parasa) [reply src/hotspot/share/opto/comparenode.hpp:67]: Yes, that is the intention.

File: src/hotspot/cpu/x86/x86_64.ad [hunk 6]
```diff
@@ -12100,6 +12100,112 @@ instruct overflowMulL_rReg_imm(rFlagsReg cr, rRegL op1, immL32 op2, rRegL tmp)
...
+
+    Label done;
+    __ movl(tmp, -1);
+    __ cmpl(op1, op2);
+    __ jccb(Assembler::below, done);
+	  __ movl(tmp, 1);
+    __ jccb(Assembler::above, done);
+	  __ movl(tmp, 0);
+    __ bind(done);
+    __ movl(dest, tmp);
```
[the remark this answers is not in the document]
Comment (vamsi-parasa) [reply src/hotspot/cpu/x86/x86_64.ad:12178]: Sure, will do that and update it soon.

File: src/hotspot/cpu/x86/x86_64.ad [hunk 7]
```diff
@@ -12100,6 +12100,112 @@ instruct overflowMulL_rReg_imm(rFlagsReg cr, rRegL op1, immL32 op2, rRegL tmp)
...
+
+    Label done;
+    __ movl(tmp, -1);
+    __ cmpq(op1, op2);
+    __ jccb(Assembler::below, done);
+	  __ movl(tmp, 1);
+    __ jccb(Assembler::above, done);
+	  __ movl(tmp, 0);
+    __ bind(done);
+    __ movl(dest, tmp);
```
[the remark this answers is not in the document]
Comment (vamsi-parasa) [reply src/hotspot/cpu/x86/x86_64.ad:12204]: Sure, will do that and update it soon.

File: test/micro/org/openjdk/bench/java/lang/CompareInteger.java (hunk 1 above)
Comment (vamsi-parasa) [reply test/micro/org/openjdk/bench/java/lang/CompareInteger.java:78]: The suggested refactoring is definitely elegant but one rare possibility is overflow due to the addition/subtraction. The swap logic doesn't have that problem.

File: test/micro/org/openjdk/bench/java/lang/CompareLong.java (hunk 2 above)
Comment (vamsi-parasa) [reply test/micro/org/openjdk/bench/java/lang/CompareLong.java:5]: Sure, will do the unification.

File: src/hotspot/cpu/x86/x86_64.ad (hunk 4 above)
Comment (vamsi-parasa) [reply src/hotspot/cpu/x86/x86_64.ad:12125]: Sure, will refactor it into a macro-assembly

**Comment by merykitty** [comment]
@vamsi-parasa  But normally the result of the `compare` methods is not used as a raw integer (it is useless to do so since the methods do not have any promise regarding the value of the result, only its sign). The idiom is to compare the result with 0, such as `Integer.compare(x, y) > 0`, the compiler can reduce this to `x > y` (last time I checked it does not do so but in principle this is possible). Your intrinsic prevents the compiler to do such transformations, hurting the performance of real programs.

> Even without the intrinsic, the compiler is currently able to reduce it to x u< y

It is because the compiler can recognise the pattern `x + MIN_VALUE < y + MIN_VALUE` and transforms it into `x u< y`. This transformation is fragile however if one of the arguments is in the form `x + con`, in such cases constant propagation may lead to slight deviations from recognised patterns, defeat the transformations. As a result, it may be justifiable to have a dedicated intrinsic for that since unsigned comparisons are pretty basic operations that are needed for optimal range check performance.

Thanks.

File: test/micro/org/openjdk/bench/java/lang/CompareInteger.java (hunk 1 above)
Comment (jatin-bhateja) [reply test/micro/org/openjdk/bench/java/lang/CompareInteger.java:78]: > The suggested refactoring is definitely elegant but one rare possibility is overflow due to the addition/subtraction. The swap logic doesn't have that problem.

Given that BUFFER_SIZE is 1024 can we not circumvent this problem by using rng.nextLong(4096) in initialization loop.

**Comment by jatin-bhateja** [comment]
> @vamsi-parasa But normally the result of the `compare` methods is not used as a raw integer (it is useless to do so since the methods do not have any promise regarding the value of the result, only its sign). The idiom is to compare the result with 0, such as `Integer.compare(x, y) > 0`, the compiler can reduce this to `x > y` (last time I checked it does not do so but in principle this is possible). Your intrinsic prevents the compiler from doing such transformations, hurting the performance of real programs.
> 
> > Even without the intrinsic, the compiler is currently able to reduce it to x u< y
> 
> It is because the compiler can recognise the pattern `x + MIN_VALUE < y + MIN_VALUE` and transforms it into `x u< y`. This transformation is fragile however if one of the arguments is in the form `x + con`, in such cases constant propagation may lead to slight deviations from recognised patterns, defeat the transformations. As a result, it may be justifiable to have a dedicated intrinsic for that since unsigned comparisons are pretty basic operations that are needed for optimal range check performance.
> 
> Thanks.

I agree, what @merykitty is proposing its to add another transformation on top of newly generated IR nodes which can reduce the emitted JIT sequence,  along with this constant operand handling for newly created IR nodes though Value routines will also be useful.

**Pull request closed without merging**